### DISCUSSION TITLE
protocolServicesSupported gives writeAccessDenied exception

### DISCUSSION THREAD

**Discussion by timcurry1** [JoelBender/bacpypes#457] [Q&A]
I am the developer using the bacpypes. Someone on the network stated that my protocolServicesSupported is null. He confirmed it using wireshark. I am surprised that he needed that attribute because no one else has ever needed it. All of my analog-Inputs are RandomAnalogValueObject as you can see in the attachment pdf.

I can run the following code in the __init__ method:
        services_supported = this_application.get_services_supported()
        print("services_supported:")
        print(services_supported)
        print(services_supported.value)
        
        this_device.protocolServicesSupported = services_supported.value

but when I try to assign services_supported.value to this_device.protocolServicesSupported, I get a writeAccessDenied exception.
Should bacpypes automatically assign this attribute and place it on the network? Not sure why I have to do this for him.

He also wants protocolObjectTypeSupported. I have not looked into this yet.

I have YABE on my Window laptop and everything appears to run fine. 
I am using bacpypes.__version__ = 1.18.1. with python 2.7. on a Linux computer running Debian.

[BACnetService.pdf](https://github.com/JoelBender/bacpypes/files/8695688/BACnetService.pdf)

**Comment by JoelBender** [comment]
The `protocolServicesSupported` property is read-only, defined in `bacpypes.local.device`, and computed based on introspection of the application instance (the `get_services_supported()` method of the `Application` class). Many clients use this property to determine if it can use ReadPropertyMultiple rather than ReadProperty.  The Python2.7 part of the library has not kept up-to-date with the other versions, but this should be working as of version `0.18.5`.  Please update and try again.

The `protocolObjectTypesSupported` has not been implemented, in theory it would be a bit string of the types of instances that are created (clearly if a device has an instance of a particular type of object it supports that type) but the Clause 12.11.15 in the standard states:

> This property, of type BACnetObjectTypesSupported, indicates which standardized object types can be present in this device's
protocol implementation.

And since all BACnet objects _could_ be present, but BACpypes does not support much of the required behavior of many of the types, so any implementation using them would be particularly lame.

**Reply by Anthony-J-Garot** [reply]
Hi

> The protocolServicesSupported property is read-only, defined in bacpypes.local.device, and computed based on introspection of the application instance (the get_services_supported() method of the Application class). 

But the documentation here suggests otherwise:

https://bacpypes.readthedocs.io/en/latest/samples/sample001.html

```
.
.
.
# let the device object know
this_device.protocolServicesSupported = services_supported.value
```

The code isn't actually in the example . . . I added it from the doc. Is the doc outdated?

**Reply by JoelBender** [reply]
Ugh!   Yes, the documentation hasn't been updated in ages.

**Comment by timcurry1** [comment]
Thanks Joel,
I downloaded the 0.18.5. Is there a way to place print statements in the callback (or network reader) whenever this guys sends a request for things like protocolRevision, firmwareRevision, and servicesSupported? Can you tell me what python file handles this callback (or network reader). Most of these request work except for a few.

**Comment by JoelBender** [comment]
The simplest way to print out what function calls are being made and the contents of their parameters is to use the debugging tool, like adding `--debug bacpypes.app` will attach a log handler to the module, or `--debug bacpypes.app.Application` will attach a log handler to the class.  In addition to the function call, you will also get a few "progress messages" with key statements on what is being decoded, I essentially stuffed in a bunch of print statements while developing the code and left them in as logging calls.

Adding `--color` is a nice way to visually separate different logging handlers, I quite often use combinations like `--debug __main__ bacpypes.bvllservice.AnnexJCodec --color`.

You can redirect the output of a logger to a file like `--debug bacpypes.object:somefile.txt` which is handy for separating specific "layers" of code into different files, and for long running processes add `--debug bacpypes.netservice:somefile.txt:1048576:5` to keep the last five 10MB rolling log files.  Check out the `bacpypes.consolelogging` module for details.

**Reply by timcurry1** [reply]
Thanks again Joel,
I want to get the debug logger working.
I added your suggestion to my code:
        parser.add_argument('--debug bacpypes.app')
        parser.add_argument('--debug bacpypes.app.Application')

right before:
args = parser.parse_args()

I also set the _debug boolean to 1 at the top of my code. My code is a thread that runs.
Where does the debug information go? stdout? stderr? and log file? Maybe the add_argument() is not the correct way to run debugger.

Sorry, I read the docs but I am still confused.

My code was attached above prior to these changes called BACnetService.pdf.

**Reply by timcurry1** [reply]
Joel, 
I was able to use my BACnet service without the thread so I can now add the --debug on the Linux commands line. I still would like your suggestions getting the --debug XXXXx using the parser.add_argument function.

I used:
./BACnetService.py --debug bacpypes.app.Application --color
and I noticed some interesting errors when I send ReadPropertyRequest using YABE. YABE even prints an error in the log widget.

Should I be over writing the RandomValueProperty::ReadProperty? The following error looks like I don't have the property it is requesting. I am confused.

    <bacpypes.apdu.ReadPropertyRequest(12,182) instance at 0x76560430>
        pduSource = <Address 0xc0a85621cb16>
        pduExpectingReply = True
        pduNetworkPriority = 0
        apduType = 0
        apduSeg = False
        apduMor = False
        apduSA = True
        apduMaxSegs = 7
        apduMaxResp = 5
        apduService = 12
        apduInvokeID = 182
        objectIdentifier = ('analogValue', 0)
        propertyIdentifier = 'propertyList'
        pduData = x''^[[0m
^[[32mDEBUG:bacpypes.app.Application:    - helperFn: <bound method BIPSimpleApplication.do_ReadPropertyRequest of <bacpypes.app.BIPSimpleApplication object at 0x76548f50>>^[[0m
^[[32mDEBUG:bacpypes.app.Application:    - execution error: ExecutionError('property', 'unknownProperty')^[[0m
^[[32mDEBUG:bacpypes.app.Application:indication <bacpypes.apdu.ReadPropertyRequest(12,183) instance at 0x76560730>


I redirected the output of debug to bac3.log and included it. 

[bac3.log](https://github.com/JoelBender/bacpypes/files/8711242/bac3.log)

**Comment by JoelBender** [comment]
Please add this to the other imports you have:
```python
from bacpypes.local.object import CurrentPropertyListMixIn
```
and then include the mixin class where you have defined your class RandomAnalogValueObject:
```python
class RandomAnalogValueObject(CurrentPropertyListMixIn, AnalogValueObject):
    ...
```
That should fix it.

In the `bacpypes.consolelogging` module is a function `ConsoleLogHandler()` which you pass the logger object directly or the name of the logger.  The next is an instance of `logging.StreamHandler` (the built-in Python logging module that directs content to `stderr`) or a subclass of it (like logging.handlers.RotatingFileHandler), optionally the logging level (usually debug) and an integer for color (checkout the `LoggingFormatter` class in the debugging module).

The `ConsoleLogHandler()` function will take care of setting the appropriate `_debug` switch in the module as necessary, so you can do simple things like call `ConsoleLogHandler("__main__")` and get much of the same functionality without changing the argument parser.

**Reply by timcurry1** [reply]
That fixed my errors, thanks.

Where do I place ConsoleLogHandler("__main__")? I tried placing it in bacpypes/consolelogging.py but it threw an exception:
not a valid logger name: '__main__'

I want to turn on enough debugging to understand why the client gets an error when requesting protocolServicesSupported.

**Comment by JoelBender** [comment]
When you run a Python application the `__name__` is the string `"__main__"` with double-underscores, so that's what you have to pass to the function, and you make that call in your application after you have imported BACpypes and created the module level logging handler, just like the sample applications:
```python
# some debugging
_debug = 0
_log = ModuleLogger(globals())
```
You can also call it with the name of a BACpypes module or class, like `ConsoleLogHandler("bacpypes.app.Application")`.

**Comment by timcurry1** [comment]
Joel,
I asked my BACnet client counterpart person to put together some notes on his commands as I am getting my bacpypes errors. I also have the bacpypes logs in the next reply message so I don't confuse his discussion. Here are his notes:

We are trying to bring the BACPYPES device into a Tridium Niagara JACE 8000
When we discover the device, it holds all necessary properties (sw version, etc)
When we add the device to the niagara database – it immediately goes “offline/down”
When I try to manually “ping” the device via niagara, it seems to cause an error on the BACPYPES device and it remains “down”

Because of its down status, we are not able to attempt a discover of other objects in the device (analog values etc)
Wireshark shows the request for system-status here and the response “error” from the BACPYPES device

**Comment by timcurry1** [comment]
Joel,
I included the ConsoleLogHandlers:
ConsoleLogHandler( "bacpypes.app") 
ConsoleLogHandler( "bacpypes.bvllservice.AnnexJCodec") 
ConsoleLogHandler( "bacpypes.app.Application") 
at the top of my program.

I received a log error when my client buddy sent the 'add' to database. The pdf file is attached. I get an additional error when he pings my BACnet server. All other informational logs seem to indicate that everything else is fine.

[journal_BACnet.pdf](https://github.com/JoelBender/bacpypes/files/8743325/journal_BACnet.pdf)

**Comment by JoelBender** [comment]
It's looking for device object properties that you haven't provided.  Starting at packet 126 you can see a series of ReadPropertyRequest PDUs coming in for objectName, objectList, protocolServicesSupported, vendorName, modelName, protocolRevision, firmwareRevision, applicationSoftwareVersion, and the critical one is probably systemStatus.  Give these properties values when you create your device object and try again.

The error at packet 578 implies they are asked for a property where BACpypes seems to be having trouble decoding what they sent.  If you can attach a Wireshark capture file I can see the rest of the layers, otherwise text files are easier than PDFs.

**Reply by timcurry1** [reply]
I'm trying to send the Wireshark file with the client's UDP request. I cannot seem to attach/drop the *.pcapng file to this discussion group. I had the same problem with other suffixes so I made PDF out of them. Can I email them to you? My email is in my profile.

**Reply by timcurry1** [reply]
Joel,
It is my assumption that the protocolServicesSupported and systemStatus are set by BACpypes and read only. Your first email above confirmed this. Is there a way in BACpypes to print it out at the point where the client requests these parameters?

**Reply by timcurry1** [reply]
Hey Joel,

Do you any thoughts on our replies above?

**Reply by JoelBender** [reply]
I sent you an email, check your gmail inbox.

Attaching a debugger at the UDP layer will dump out the contents of incoming packets, but they won't re recognized as that until they have been through the BVLL, network, and up to the application layer.  You probably want to attach a debugger like `--debug bacpypes.service.object.ReadWritePropertyServices` which is just the mix-in class for handling Read Property and Write Property services.

**Reply by timcurry1** [reply]
Joel,
I found your email in my junk folder. I made a contact so that should not happen again. Sorry for the confusion.

We did more testing after included the new debugger command as indicated above. I copied the interesting section. I also included the file as an attachment. It appears bacpypes did not like the value of 0 for system status.


Jun 02 15:37:29 imx6ul-var-dart which.py[279]: DEBUG:bacpypes.app.Application:indication <bacpypes.apdu.ReadPropertyRequest(12,161) instance at 0x751c1dd0>
Jun 02 15:37:29 imx6ul-var-dart which.py[279]:     <bacpypes.apdu.ReadPropertyRequest(12,161) instance at 0x751c1dd0>
Jun 02 15:37:29 imx6ul-var-dart which.py[279]:         pduSource = <Address 10.10.19.181>
Jun 02 15:37:29 imx6ul-var-dart which.py[279]:         pduExpectingReply = True
Jun 02 15:37:29 imx6ul-var-dart which.py[279]:         pduNetworkPriority = 0
Jun 02 15:37:29 imx6ul-var-dart which.py[279]:         apduType = 0
Jun 02 15:37:29 imx6ul-var-dart which.py[279]:         apduSeg = False
Jun 02 15:37:29 imx6ul-var-dart which.py[279]:         apduMor = False
Jun 02 15:37:29 imx6ul-var-dart which.py[279]:         apduSA = True
Jun 02 15:37:29 imx6ul-var-dart whi
[foo_jour.log](https://github.com/JoelBender/bacpypes/files/8825513/foo_jour.log)
ch.py[279]:         apduMaxSegs = 0
Jun 02 15:37:29 imx6ul-var-dart which.py[279]:         apduMaxResp = 3
Jun 02 15:37:29 imx6ul-var-dart which.py[279]:         apduService = 12
Jun 02 15:37:29 imx6ul-var-dart which.py[279]:         apduInvokeID = 161
Jun 02 15:37:29 imx6ul-var-dart which.py[279]:         objectIdentifier = ('device', 599)
Jun 02 15:37:29 imx6ul-var-dart which.py[279]:         propertyIdentifier = 'systemStatus'
Jun 02 15:37:29 imx6ul-var-dart which.py[279]:         pduData = x''
Jun 02 15:37:29 imx6ul-var-dart which.py[279]: DEBUG:bacpypes.app.Application:    - helperFn: <bound method BIPSimpleApplication.do_ReadPropertyRequest of <bacpypes.app.BIPSimpleApplication object at 0x751acab0>>
Jun 02 15:37:29 imx6ul-var-dart which.py[279]: DEBUG:bacpypes.service.object.ReadWritePropertyServices:do_ReadPropertyRequest <bacpypes.apdu.ReadPropertyRequest(12,161) instance at 0x751c1dd0>
Jun 02 15:37:29 imx6ul-var-dart which.py[279]:     <bacpypes.apdu.ReadPropertyRequest(12,161) instance at 0x751c1dd0>
Jun 02 15:37:29 imx6ul-var-dart which.py[279]:         pduSource = <Address 10.10.19.181>
Jun 02 15:37:29 imx6ul-var-dart which.py[279]:         pduExpectingReply = True
Jun 02 15:37:29 imx6ul-var-dart which.py[279]:         pduNetworkPriority = 0
Jun 02 15:37:29 imx6ul-var-dart which.py[279]:         apduType = 0
Jun 02 15:37:29 imx6ul-var-dart which.py[279]:         apduSeg = False
Jun 02 15:37:29 imx6ul-var-dart which.py[279]:         apduMor = False
Jun 02 15:37:29 imx6ul-var-dart which.py[279]:         apduSA = True
Jun 02 15:37:29 imx6ul-var-dart which.py[279]:         apduMaxSegs = 0
Jun 02 15:37:29 imx6ul-var-dart which.py[279]:         apduMaxResp = 3
Jun 02 15:37:29 imx6ul-var-dart which.py[279]:         apduService = 12
Jun 02 15:37:29 imx6ul-var-dart which.py[279]:         apduInvokeID = 161
Jun 02 15:37:29 imx6ul-var-dart which.py[279]:         objectIdentifier = ('device', 599)
Jun 02 15:37:29 imx6ul-var-dart which.py[279]:         propertyIdentifier = 'systemStatus'
Jun 02 15:37:29 imx6ul-var-dart which.py[279]:         pduData = x''
Jun 02 15:37:29 imx6ul-var-dart which.py[279]: DEBUG:bacpypes.service.object.ReadWritePropertyServices:    - object: <bacpypes.local.device.LocalDeviceObject object at 0x76cbe6b0>
Jun 02 15:37:29 imx6ul-var-dart which.py[279]: DEBUG:bacpypes.service.object.ReadWritePropertyServices:    - datatype: <class 'bacpypes.basetypes.DeviceStatus'>
Jun 02 15:37:29 imx6ul-var-dart which.py[279]: DEBUG:bacpypes.service.object.ReadWritePropertyServices:    - value: '0'
Jun 02 15:37:29 imx6ul-var-dart which.py[279]: ERROR:bacpypes.app.Application:exception: ValueError("undefined enumeration '0'",)
Jun 02 15:37:29 imx6ul-var-dart which.py[279]: Traceback (most recent call last):
Jun 02 15:37:29 imx6ul-var-dart which.py[279]:   File "/home/user/python/EC2/bacpypes/app.py", line 382, in indication
Jun 02 15:37:29 imx6ul-var-dart which.py[279]:     helperFn(apdu)
Jun 02 15:37:29 imx6ul-var-dart which.py[279]:   File "/home/user/python/EC2/bacpypes/service/object.py", line 66, in do_ReadPropertyRequest
Jun 02 15:37:29 imx6ul-var-dart which.py[279]:     value = datatype(value)
Jun 02 15:37:29 imx6ul-var-dart which.py[279]:   File "/home/user/python/EC2/bacpypes/primitivedata.py", line 1150, in __init__
Jun 02 15:37:29 imx6ul-var-dart which.py[279]:     raise ValueError("undefined enumeration '%s'" % (arg,))
Jun 02 15:37:29 imx6ul-var-dart which.py[279]: ValueError: undefined enumeration '0'
Jun 02 15:37:29 imx6ul-var-dart which.py[279]: ERROR:bacpypes.app.Application:exception: ValueError("undefined enumeration '0'",)
Jun 02 15:37:29 imx6ul-var-dart which.py[279]: Traceback (most recent call last):
Jun 02 15:37:29 imx6ul-var-dart which.py[279]:   File "/home/user/python/EC2/bacpypes/app.py", line 382, in indication
Jun 02 15:37:29 imx6ul-var-dart which.py[279]:     helperFn(apdu)
Jun 02 15:37:29 imx6ul-var-dart which.py[279]:   File "/home/user/python/EC2/bacpypes/service/object.py", line 66, in do_ReadPropertyRequest
Jun 02 15:37:29 imx6ul-var-dart which.py[279]:     value = datatype(value)
Jun 02 15:37:29 imx6ul-var-dart which.py[279]:   File "/home/user/python/EC2/bacpypes/primitivedata.py", line 1150, in __init__
Jun 02 15:37:29 imx6ul-var-dart which.py[279]:     raise ValueError("undefined enumeration '%s'" % (arg,))
Jun 02 15:37:29 imx6ul-var-dart which.py[279]: ValueError: undefined enumeration '0'
Jun 02 15:37:29 imx6ul-var-dart which.py[279]: DEBUG:bacpypes.bvllservice.AnnexJCodec:indication <bacpypes.bvll.OriginalUnicastNPDU object at 0x751ac4b0>
Jun 02 15:37:29 imx6ul-var-dart which.py[279]:     <bacpypes.bvll.OriginalUnicastNPDU object at 0x751ac4b0>
Jun 02 15:37:29 imx6ul-var-dart which.py[279]:         pduDestination = <Address 10.10.19.181>
Jun 02 15:37:29 imx6ul-var-dart which.py[279]:         pduExpectingReply = 0
Jun 02 15:37:29 imx6ul-var-dart which.py[279]:         pduNetworkPriority = 0
Jun 02 15:37:29 imx6ul-var-dart which.py[279]:         bvlciType = 129
Jun 02 15:37:29 imx6ul-var-dart which.py[279]:         bvlciFunction = 10
Jun 02 15:37:29 imx6ul-var-dart which.py[279]:         bvlciLength = 13
Jun 02 15:37:29 imx6ul-var-dart which.py[279]:         pduData = x'01.00.50.a1.0c.91.00.91.19'

**Comment by JoelBender** [comment]
The `systemStatus` property (see Clause **12.11.4 System_Status**) is an enumeration of type DeviceStatus (called BACnetDeviceStatus in the standard, defined in the `bacpypes.basetypes` module) so you need to provide the value `"operational"`.  Apparently you passed in the string "0" rather than the integer 0 (maybe from an INI file?) so BACpypes was attempting to convert the name to a value.

**Reply by timcurry1** [reply]
That fixed all of our problems.  Thanks so much for all of your help and thanks for providing a great python package to our community.
